api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,275 @@
+
+  // Add or replace HTTP headers. This will be ignored after SEND_XXX_BODY_CHUNK has
+  // been sent. If SEND_TRAILERS was previously sent, then
+  // modify the trailers instead.
+  repeated config.core.v3.HeaderValueOption set_headers = 3;
+
+  // Remove these HTTP headers. This will be ignored after SEND_XXX_BODY_CHUNK
+  // has been sent. If SEND_XXX_TRAILERS was previously sent, then
+  // modify the trailers instead.
+  repeated string remove_headers = 4;
+
+  // Replace some or all of the last message body received with a new
+  // set of bytes. This will be ignored unless the previous message sent
+  // by the client is a "XXXBODY" request.
+  BodyMutation body_mutation = 5;
+
+  // Add new metadata to the Envoy context
+  google.protobuf.Struct new_metadata = 6;
+
+  // Change how and if the filter waits for responses from the server
+  ProcessingModeChange async_mode_change = 7;
+}
+
+// This message is sent to the mutation server when the HTTP request is
+// first received.
+// [#next-free-field: 9]
+message Request {
+  // A unique ID that will match the one sent in the "Response" message
+  string id = 1;
+
+  // The HTTP request method, such as `GET`, `POST`.
+  string method = 2;
+
+  // The HTTP request headers. If multiple headers share the same key, they
+  // must be merged according to the HTTP spec. All header keys must be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 3;
+
+  // The request target, as it appears in the first line of the HTTP request. This includes
+  // the URL path and query-string. No decoding is performed.
+  string path = 4;
+
+  // The HTTP request `Host` or 'Authority` header value.
+  string host = 5;
+
+  // The HTTP URL scheme, such as `http` and `https`.
+  string scheme = 6;
+
+  // Information about the upstream service's connection
+  auth.v3.AttributeContext.Peer client_info = 7;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 8;
+}
+
+// This message is sent to the mutation server when the HTTP response is first
+// received.
+// [#next-free-field: 6]
+message Response {
+  // A unique ID that will match the one sent in the "Request" message
+  string id = 1;
+
+  // The HTTP status on the response
+  int32 status_code = 2;
+
+  // The HTTP headers
+  config.core.v3.HeaderMap headers = 3;
+
+  // Information about the upstream service's connection
+  auth.v3.AttributeContext.Peer destination_info = 4;
+
+  // The values of properties selected by the "response_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 5;
+}
+
+// This message contains the message body that Envoy sends to the mutation service.
+message Body {
+  bytes body = 1;
```

Comment:
This could probably be a vector of body chunks, otherwise Envoy will have to linearize it before sending to the external service.

Reply:
+1

Reply:
I was planning on using the "StopIterationAndBuffer" feature, which causes Envoy to do just this, but I can see the point.

However, if we look above to the "BodyMutation" proto, it changes the message body based on the size of the last chunk that we sent the server. So in that case, either we need a way to separately mutate each chunk (ick) or we tell the user that the BodyMutation is relative to the list of chunks shown here assuming that they are all concatenated together.

So I think this is a matter of making things easier to implement on the server side versus making Envoy as efficient as we can.